File: scripts/prepare_data.py
from __future__ import annotations

from pathlib import Path
import json
import numpy as np
import pandas as pd

PROJECT = Path('/home/ubuntu/Estatistica-escolha-dos-jogadores-copa-do-mundo-2026')
RAW_FILE = PROJECT / 'data' / 'raw' / 'players_data_light-2025_2026.csv'
PROCESSED_DIR = PROJECT / 'data' / 'processed'
DOCS_DIR = PROJECT / 'docs'
PROCESSED_DIR.mkdir(parents=True, exist_ok=True)
DOCS_DIR.mkdir(parents=True, exist_ok=True)

SOURCE_METADATA = {
    'dataset_title': 'Football Players Stats (2025-2026)',
    'dataset_url': 'https://www.kaggle.com/datasets/hubertsidorowicz/football-players-stats-2025-2026',
    'creator': 'Hubert Sidorowicz',
    'platform': 'Kaggle',
    'original_source': 'FBref',
    'season': '2025-2026',
    'version': 34,
    'license': 'MIT',
    'author_of_project': 'Diogo Rego - Estudante de Estatística UFPB',
}

NUMERIC_COLUMNS = [
    'Age', 'MP', 'Starts', 'Min', '90s', 'Gls', 'Ast', 'G+A', 'G-PK', 'PK', 'PKatt',
    'CrdY', 'CrdR', 'G+A-PK', 'Sh', 'SoT', 'SoT%', 'Sh/90', 'SoT/90', 'G/Sh',
    'G/SoT', 'Crs', 'TklW', 'Int', 'Fld', 'Fls', 'OG', 'GA', 'GA90', 'SoTA',
    'Saves', 'Save%', 'W', 'D', 'L', 'CS', 'CS%', 'PKA', 'PKsv', 'PKm'
]


def normalize_nation(value: object) -> str:
    if pd.isna(value):
        return 'Não informado'
    parts = str(value).split()
    return parts[-1] if parts else str(value)


def primary_position(pos: object) -> str:
    if pd.isna(pos):
        return 'OUT'
    text = str(pos).upper()
    if 'GK' in text:
        return 'GK'
    if 'DF' in text:
        return 'DF'
    if 'MF' in text:
        return 'MF'
    if 'FW' in text:
        return 'FW'
    return 'OUT'


def per90(series: pd.Series, ninety: pd.Series) -> pd.Series:
    safe_90 = ninety.replace(0, np.nan)
    return (series / safe_90).replace([np.inf, -np.inf], np.nan).fillna(0)


def minmax_by_position(df: pd.DataFrame, column: str) -> pd.Series:
    grouped = df.groupby('posicao_principal')[column]
    min_v = grouped.transform('min')
    max_v = grouped.transform('max')
    spread = (max_v - min_v).replace(0, np.nan)
    return ((df[column] - min_v) / spread).fillna(0)


def build_dataset() -> pd.DataFrame:
    df = pd.read_csv(RAW_FILE)

    for col in NUMERIC_COLUMNS:
        if col in df.columns:
            df[col] = pd.to_numeric(df[col], errors='coerce')

    df['jogador'] = df['Player'].astype(str).str.strip()
    df['nacao_codigo'] = df['Nation'].apply(normalize_nation)
    df['posicao_original'] = df['Pos'].fillna('Não informado').astype(str)
    df['posicao_principal'] = df['Pos'].apply(primary_position)
    df['clube'] = df['Squad'].fillna('Não informado').astype(str)
    df['liga'] = df['Comp'].fillna('Não informado').astype(str)
    df['idade'] = df['Age'].fillna(df['Age'].median()).round(1)
    df['minutos'] = df['Min'].fillna(0)
    df['jogos'] = df['MP'].fillna(0)
    df['noventas'] = df['90s'].fillna(0).clip(lower=0)

    raw_components = {
        'gols_p90': 'Gls',
        'assistencias_p90': 'Ast',
        'participacoes_gol_p90': 'G+A',
        'finalizacoes_p90': 'Sh',
        'finalizacoes_gol_p90': 'SoT',
        'cruzamentos_p90': 'Crs',
        'desarmes_vencidos_p90': 'TklW',
        'interceptacoes_p90': 'Int',
        'faltas_sofridas_p90': 'Fld',
    }

    for new_col, old_col in raw_components.items():
        df[new_col] = per90(df[old_col].fillna(0), df['noventas'])

    df['disciplina_p90'] = per90(df['CrdY'].fillna(0) + 2 * df['CrdR'].fillna(0), df['noventas'])
    df['eficiencia_chute'] = df['G/Sh'].fillna(0).clip(lower=0, upper=1)
    df['taxa_chutes_gol'] = (df['SoT%'].fillna(0) / 100).clip(lower=0, upper=1)
    df['save_pct'] = (df['Save%'].fillna(0) / 100).clip(lower=0, upper=1)
    df['clean_sheet_pct'] = (df['CS%'].fillna(0) / 100).clip(lower=0, upper=1)
    df['gols_sofridos_inv'] = -df['GA90'].fillna(df['GA90'].median()).fillna(0)

    for col in [
        'gols_p90', 'assistencias_p90', 'participacoes_gol_p90', 'finalizacoes_p90',
        'finalizacoes_gol_p90', 'cruzamentos_p90', 'desarmes_vencidos_p90',
        'interceptacoes_p90', 'faltas_sofridas_p90', 'eficiencia_chute', 'taxa_chutes_gol',
        'save_pct', 'clean_sheet_pct', 'gols_sofridos_inv'
    ]:
        df[f'{col}_norm'] = minmax_by_position(df, col)

    df['volume_jogo_norm'] = np.log1p(df['minutos']) / np.log1p(max(df['minutos'].max(), 1))
    df['experiencia_norm'] = np.log1p(df['jogos']) / np.log1p(max(df['jogos'].max(), 1))
    df['idade_prime_norm'] = np.exp(-((df['idade'] - 27) ** 2) / (2 * 4.5 ** 2))

    attack_score = (
        0.30 * df['participacoes_gol_p90_norm'] +
        0.18 * df['gols_p90_norm'] +
        0.16 * df['assistencias_p90_norm'] +
        0.12 * df['finalizacoes_gol_p90_norm'] +
        0.10 * df['eficiencia_chute_norm'] +
        0.08 * df['cruzamentos_p90_norm'] +
        0.06 * df['volume_jogo_norm']
    )
    midfield_score = (
        0.22 * df['assistencias_p90_norm'] +
        0.18 * df['cruzamentos_p90_norm'] +
        0.16 * df['faltas_sofridas_p90_norm'] +
        0.14 * df['participacoes_gol_p90_norm'] +
        0.12 * df['desarmes_vencidos_p90_norm'] +
        0.10 * df['interceptacoes_p90_norm'] +
        0.08 * df['volume_jogo_norm']
    )
    defense_score = (
        0.30 * df['desarmes_vencidos_p90_norm'] +
        0.26 * df['interceptacoes_p90_norm'] +
        0.12 * df['cruzamentos_p90_norm'] +
        0.10 * df['volume_jogo_norm'] +
        0.10 * df['experiencia_norm'] +
        0.07 * df['idade_prime_norm'] +
        0.05 * (1 - df['disciplina_p90'].clip(0, 1))
    )
    goalkeeper_score = (
        0.42 * df['save_pct_norm'] +
        0.24 * df['clean_sheet_pct_norm'] +
        0.16 * df['gols_sofridos_inv_norm'] +
        0.10 * df['volume_jogo_norm'] +
        0.08 * df['experiencia_norm']
    )

    df['score_ataque'] = attack_score
    df['score_meio'] = midfield_score
    df['score_defesa'] = defense_score
    df['score_goleiro'] = goalkeeper_score

    df['indice_desempenho_base'] = np.select(
        [df['posicao_principal'].eq('GK'), df['posicao_principal'].eq('DF'), df['posicao_principal'].eq('MF'), df['posicao_principal'].eq('FW')],
        [df['score_goleiro'], df['score_defesa'], df['score_meio'], df['score_ataque']],
        default=(attack_score + midfield_score + defense_score) / 3,
    )

    df['penalidade_baixa_amostra'] = np.where(df['minutos'] >= 900, 1, np.sqrt(df['minutos'].clip(lower=1) / 900))
    df['indice_desempenho'] = (100 * df['indice_desempenho_base'] * df['penalidade_baixa_amostra']).clip(0, 100)
    df['incerteza_modelo'] = (18 / np.sqrt(df['noventas'].clip(lower=1)) + 4 * (1 - df['volume_jogo_norm'])).clip(3, 28)

    selected_columns = [
        'jogador', 'nacao_codigo', 'posicao_original', 'posicao_principal', 'clube', 'liga',
        'idade', 'jogos', 'minutos', 'noventas', 'Gls', 'Ast', 'G+A', 'Sh', 'SoT', 'SoT%',
        'Crs', 'TklW', 'Int', 'GA90', 'Saves', 'Save%', 'CS%', 'gols_p90', 'assistencias_p90',
        'participacoes_gol_p90', 'finalizacoes_p90', 'finalizacoes_gol_p90', 'cruzamentos_p90',
        'desarmes_vencidos_p90', 'interceptacoes_p90', 'save_pct', 'clean_sheet_pct',
        'score_ataque', 'score_meio', 'score_defesa', 'score_goleiro', 'indice_desempenho',
        'incerteza_modelo', 'penalidade_baixa_amostra'
    ]

    out = df[selected_columns].copy()
    out = out.sort_values(['indice_desempenho', 'minutos'], ascending=[False, False])
    return out


def main() -> None:
    data = build_dataset()
    output_csv = PROCESSED_DIR / 'players_world_cup_monte_carlo_ready.csv'
    data.to_csv(output_csv, index=False)

    summary = {
        **SOURCE_METADATA,
        'rows': int(len(data)),
        'countries': int(data['nacao_codigo'].nunique()),
        'clubs': int(data['clube'].nunique()),
        'leagues': int(data['liga'].nunique()),
        'processed_file': str(output_csv.relative_to(PROJECT)),
    }

    with open(PROCESSED_DIR / 'metadata.json', 'w', encoding='utf-8') as f:
        json.dump(summary, f, ensure_ascii=False, indent=2)

    top = data.head(30)[['jogador', 'nacao_codigo', 'posicao_principal', 'clube', 'liga', 'minutos', 'indice_desempenho']]
    top.to_csv(PROCESSED_DIR / 'top30_indice_desempenho.csv', index=False)
    print(json.dumps(summary, ensure_ascii=False, indent=2))


if __name__ == '__main__':
    main()

File: scripts/prepare_data.R
# -----------------------------------------------------------------------------
# Projeto: Simulação de Monte Carlo para Escolha Estatística de Jogadores
# Autor: Diogo Rego - Estudante de Estatística UFPB
# Fonte: Kaggle / FBref - Football Players Stats (2025-2026)
# Objetivo: preparar a base de jogadores para o dashboard Shiny.
# -----------------------------------------------------------------------------

pacotes <- c("dplyr", "readr", "stringr", "jsonlite")
pacotes_ausentes <- pacotes[!vapply(pacotes, requireNamespace, logical(1), quietly = TRUE)]

if (length(pacotes_ausentes) > 0) {
  stop(
    sprintf(
      "Instale os pacotes antes de executar este script: install.packages(c(%s))",
      paste(sprintf('"%s"', pacotes_ausentes), collapse = ", ")
    ),
    call. = FALSE
  )
}

library(dplyr)
library(readr)
library(stringr)
library(jsonlite)

arquivo_entrada <- "data/raw/players_data_light-2025_2026.csv"
diretorio_saida <- "data/processed"
dir.create(diretorio_saida, recursive = TRUE, showWarnings = FALSE)

normalizar_nacao <- function(x) {
  ifelse(is.na(x), "Não informado", stringr::word(as.character(x), -1))
}

posicao_principal <- function(posicao) {
  posicao <- toupper(as.character(posicao))
  dplyr::case_when(
    stringr::str_detect(posicao, "GK") ~ "GK",
    stringr::str_detect(posicao, "DF") ~ "DF",
    stringr::str_detect(posicao, "MF") ~ "MF",
    stringr::str_detect(posicao, "FW") ~ "FW",
    TRUE ~ "OUT"
  )
}

por_90 <- function(valor, noventas) {
  ifelse(is.na(noventas) | noventas <= 0, 0, dplyr::coalesce(valor, 0) / noventas)
}

normalizar_por_posicao <- function(dados, variavel) {
  dados |>
    group_by(posicao_principal) |>
    mutate(
      minimo = min(.data[[variavel]], na.rm = TRUE),
      maximo = max(.data[[variavel]], na.rm = TRUE),
      normalizado = ifelse(maximo == minimo, 0, (.data[[variavel]] - minimo) / (maximo - minimo))
    ) |>
    ungroup() |>
    pull(normalizado) |>
    tidyr::replace_na(0)
}

preparar_base <- function(arquivo) {
  dados <- readr::read_csv(arquivo, show_col_types = FALSE)

  dados <- dados |>
    mutate(
      jogador = stringr::str_squish(Player),
      nacao_codigo = normalizar_nacao(Nation),
      posicao_original = coalesce(Pos, "Não informado"),
      posicao_principal = posicao_principal(Pos),
      clube = coalesce(Squad, "Não informado"),
      liga = coalesce(Comp, "Não informado"),
      idade = as.numeric(Age),
      jogos = as.numeric(MP),
      minutos = as.numeric(Min),
      noventas = pmax(as.numeric(`90s`), 0),
      gols_p90 = por_90(as.numeric(Gls), noventas),
      assistencias_p90 = por_90(as.numeric(Ast), noventas),
      participacoes_gol_p90 = por_90(as.numeric(`G+A`), noventas),
      finalizacoes_p90 = por_90(as.numeric(Sh), noventas),
      finalizacoes_gol_p90 = por_90(as.numeric(SoT), noventas),
      cruzamentos_p90 = por_90(as.numeric(Crs), noventas),
      desarmes_vencidos_p90 = por_90(as.numeric(TklW), noventas),
      interceptacoes_p90 = por_90(as.numeric(Int), noventas),
      faltas_sofridas_p90 = por_90(as.numeric(Fld), noventas),
      disciplina_p90 = por_90(as.numeric(CrdY) + 2 * as.numeric(CrdR), noventas),
      eficiencia_chute = pmin(pmax(coalesce(as.numeric(`G/Sh`), 0), 0), 1),
      taxa_chutes_gol = pmin(pmax(coalesce(as.numeric(`SoT%`), 0) / 100, 0), 1),
      save_pct = pmin(pmax(coalesce(as.numeric(`Save%`), 0) / 100, 0), 1),
      clean_sheet_pct = pmin(pmax(coalesce(as.numeric(`CS%`), 0) / 100, 0), 1),
      gols_sofridos_inv = -coalesce(as.numeric(GA90), median(as.numeric(GA90), na.rm = TRUE), 0),
      volume_jogo_norm = log1p(coalesce(minutos, 0)) / log1p(max(coalesce(minutos, 0), na.rm = TRUE)),
      experiencia_norm = log1p(coalesce(jogos, 0)) / log1p(max(coalesce(jogos, 0), na.rm = TRUE)),
      idade_prime_norm = exp(-((coalesce(idade, median(idade, na.rm = TRUE)) - 27)^2) / (2 * 4.5^2))
    )

  variaveis_norm <- c(
    "gols_p90", "assistencias_p90", "participacoes_gol_p90", "finalizacoes_p90",
    "finalizacoes_gol_p90", "cruzamentos_p90", "desarmes_vencidos_p90", "interceptacoes_p90",
    "faltas_sofridas_p90", "eficiencia_chute", "taxa_chutes_gol", "save_pct",
    "clean_sheet_pct", "gols_sofridos_inv"
  )

  for (variavel in variaveis_norm) {
    dados[[paste0(variavel, "_norm")]] <- normalizar_por_posicao(dados, variavel)
  }

  dados |>
    mutate(
      score_ataque = 0.30 * participacoes_gol_p90_norm + 0.18 * gols_p90_norm +
        0.16 * assistencias_p90_norm + 0.12 * finalizacoes_gol_p90_norm +
        0.10 * eficiencia_chute_norm + 0.08 * cruzamentos_p90_norm + 0.06 * volume_jogo_norm,
      score_meio = 0.22 * assistencias_p90_norm + 0.18 * cruzamentos_p90_norm +
        0.16 * faltas_sofridas_p90_norm + 0.14 * participacoes_gol_p90_norm +
        0.12 * desarmes_vencidos_p90_norm + 0.10 * interceptacoes_p90_norm +
        0.08 * volume_jogo_norm,
      score_defesa = 0.30 * desarmes_vencidos_p90_norm + 0.26 * interceptacoes_p90_norm +
        0.12 * cruzamentos_p90_norm + 0.10 * volume_jogo_norm +
        0.10 * experiencia_norm + 0.07 * idade_prime_norm + 0.05 * (1 - pmin(disciplina_p90, 1)),
      score_goleiro = 0.42 * save_pct_norm + 0.24 * clean_sheet_pct_norm +
        0.16 * gols_sofridos_inv_norm + 0.10 * volume_jogo_norm + 0.08 * experiencia_norm,
      indice_desempenho_base = case_when(
        posicao_principal == "GK" ~ score_goleiro,
        posicao_principal == "DF" ~ score_defesa,
        posicao_principal == "MF" ~ score_meio,
        posicao_principal == "FW" ~ score_ataque,
        TRUE ~ (score_ataque + score_meio + score_defesa) / 3
      ),
      penalidade_baixa_amostra = ifelse(minutos >= 900, 1, sqrt(pmax(minutos, 1) / 900)),
      indice_desempenho = pmin(pmax(100 * indice_desempenho_base * penalidade_baixa_amostra, 0), 100),
      incerteza_modelo = pmin(pmax(18 / sqrt(pmax(noventas, 1)) + 4 * (1 - volume_jogo_norm), 3), 28)
    ) |>
    select(
      jogador, nacao_codigo, posicao_original, posicao_principal, clube, liga,
      idade, jogos, minutos, noventas, Gls, Ast, `G+A`, Sh, SoT, `SoT%`, Crs,
      TklW, Int, GA90, Saves, `Save%`, `CS%`, gols_p90, assistencias_p90,
      participacoes_gol_p90, finalizacoes_p90, finalizacoes_gol_p90,
      cruzamentos_p90, desarmes_vencidos_p90, interceptacoes_p90, save_pct,
      clean_sheet_pct, score_ataque, score_meio, score_defesa, score_goleiro,
      indice_desempenho, incerteza_modelo, penalidade_baixa_amostra
    ) |>
    arrange(desc(indice_desempenho), desc(minutos))
}

base_processada <- preparar_base(arquivo_entrada)
readr::write_csv(base_processada, file.path(diretorio_saida, "players_world_cup_monte_carlo_ready.csv"))
readr::write_csv(
  dplyr::slice_head(base_processada, n = 30) |>
    select(jogador, nacao_codigo, posicao_principal, clube, liga, minutos, indice_desempenho),
  file.path(diretorio_saida, "top30_indice_desempenho.csv")
)

metadados <- list(
  dataset_title = "Football Players Stats (2025-2026)",
  dataset_url = "https://www.kaggle.com/datasets/hubertsidorowicz/football-players-stats-2025-2026",
  creator = "Hubert Sidorowicz",
  platform = "Kaggle",
  original_source = "FBref",
  season = "2025-2026",
  version = 34,
  license = "MIT",
  author_of_project = "Diogo Rego - Estudante de Estatística UFPB",
  rows = nrow(base_processada),
  countries = dplyr::n_distinct(base_processada$nacao_codigo),
  clubs = dplyr::n_distinct(base_processada$clube),
  leagues = dplyr::n_distinct(base_processada$liga),
  processed_file = "data/processed/players_world_cup_monte_carlo_ready.csv"
)

jsonlite::write_json(metadados, file.path(diretorio_saida, "metadata.json"), pretty = TRUE, auto_unbox = TRUE)
message("Base processada com sucesso: ", file.path(diretorio_saida, "players_world_cup_monte_carlo_ready.csv"))
